### PR TITLE
Vagrantfile: randomize the third octet of the subnet to allow for up to 255 development environments

### PR DESCRIPTION
**Pull request by dseevr**
Octet reservation is protected by locking so it’s safe to bring up many environments at the same time.

Addresses #319

### CODE REVIEW

**Comment by dseevr** [comment]
Not ready for this to merge, just wanted to get CI testing it.  Need to remove debugging info + a bit of other cleanup.

File: Vagrantfile [hunk 1]
```diff
@@ -18,11 +18,96 @@ end
...
+      raise SubnetStateFileError.new(msg.join(" "))
+    end
+  end
+
+  # no existing subnet assignment, so reserve a new one
+
+  FileUtils.touch(SUBNET_STATE_FILE)
+  FileUtils.touch(SUBNET_STATE_LOCKFILE)
+
+  f = File.open(SUBNET_STATE_LOCKFILE)
```
Comment (erikh) [inline Vagrantfile]: you're gonna hate me, but you can chuck probably 50% of this code if you just use a directory as your database with the randomized octet as filenames in it. Locking is for free that way.

Comment (dseevr) [inline Vagrantfile]: That's actually not a bad idea at all.  Thanks.

File: Vagrantfile [hunk 2]
```diff
@@ -18,17 +18,90 @@ end
...
 
 NMONS        = ENV["VMS"] || settings['vms']
-SUBNET       = settings['subnet']
 BOX          = settings['vagrant_box']
 BOX_VERSION  = settings['box_version']
 memory       = settings['memory']
 
+SUBNET_PREFIX          = settings['subnet_prefix']
+SUBNET_ASSIGNMENT_DIR  = "/tmp/volplugin_vagrant_subnets/"
+SUBNET_ASSIGNMENT_FILE = "subnet_assignment.state"
```
Comment (erikh) [inline Vagrantfile]: this file is getting created in the wrong directory sometimes. Please use the pattern on line 9 for calculating the directory.

**Comment by erikh** [comment]
LGTM minus the issue noted above, I tested this and it works great.

For the future, just mkdir + checking for EEXIST and re-rolling with rand will work. No need for the existence checks.

**Comment by erikh** [comment]
ah; that is what you're doing, disregard. :D

File: Vagrantfile (hunk 2 above)
Comment (dseevr) [inline Vagrantfile]: Good catch.  Thanks!

I'm curious:  what other directories would it be created in?  Do we invoke vagrant from some other CWD sometimes?

**Comment by dseevr** [comment]
Looks like this is ready to merge!

File: Vagrantfile (hunk 2 above)
Comment (dseevr) [inline Vagrantfile]: Ahh, never mind.  I saw the second .state file in the system-tests dir.

**Comment by erikh** [comment]
LGTM

**Pull request merged**